### PR TITLE
MRG: Add OPM example

### PR DESCRIPTION
**Pull request by larsoner**
Todo:

- [x] make example work locally
- [x] add context manager for using custom coil def
- [x] add tests and docs for context manager
- [x] complete formatting of dataset
- [x] upload dataset
- [x] add as MNE dataset
- [x] add example file to PR
- [x] update `whats_new.rst`

### CODE REVIEW

**Comment by larsoner** [comment]
Ready for review/merge from my end.

@ rzetter @ jiivana @LauriParkkonen can you take a look at the rendered example and make sure I didn't break anything:

https://9416-1301584-gh.circle-artifacts.com/0/html/auto_examples/datasets/plot_opm_data.html

Also @rzetter please check `whats_new.rst` and `datasets_index.rst` to make sure you're happy with the URL I chose and credits, etc. Let me know if there are other names (w/URLs) I should add.

**Comment by LauriParkkonen** [comment]
Looks great!

Some minor remarks:

- low-pass filter at 40 Hz is rather low for responses to electric nerve stimulation (the sharp N20 response to median nerve stimulation is less visible)  but maybe there is too much 50-Hz interference for opening the passband to e.g. 90 Hz. Rasmus can comment on this as he’s been playing with the filter.

- I would display less (use higher threshold) for the MNE on brain surface – now it’s a bit hard to see where’s the peak


Eric Larson <notifications@github.com<mailto:notifications@github.com>> kirjoitti 19.9.2018 kello 22.46:


Ready for review/merge from my end.

@ rzetter @ jiivana @LauriParkkonen<https://github.com/LauriParkkonen> can you take a look at the rendered example and make sure I didn't break anything:

https://9416-1301584-gh.circle-artifacts.com/0/html/auto_examples/datasets/plot_opm_data.html

Also @rzetter<https://github.com/rzetter> please check whats_new.rst and datasets_index.rst to make sure you're happy with the URL I chose and credits, etc. Let me know if there are other names (w/URLs) I should add.

**Comment by rzetter** [comment]
Hi, this looks great! I have some minor comments:

-The low-pass can and should be upped to e.g. 90 Hz, the 40 Hz limit is something that might have been left from my earlier scripts by mistake. A slight high-pass could also be used (0.03 Hz or 0.1 Hz), but I guess it isn't needed. However, a 50 Hz notch filter will be useful, since we typically do have some (electrical) 50 Hz noise in the OPMs.

-Agreed with Lauri regarding the color map limits in the MNE visualization. With the current limits, the entire physiologically interesting region is saturated with white. Also, the color bar in the current figure seems to have been corrupted somehow.

-Some information about the experiment paradigm could be added to the start of  the script plot_opm_data.py, e.g.:
"Electrical median nerve stimulation was delivered to the left wrist of the subject. Somatosensory evoked fields were measured using nine QuSpin SERF OPMs placed over the right-hand side somatomotor area. Here we demonstrate how to localize these custom OPM data in MNE."

-I also sent Neuromag SQUID data for the same experiment, do you want to include that in the example for comparison purposes?

-Great that you've added a mechanism to use custom coil definitions on the fly, that will definitely be useful!

-Also great to see the tweaks to the 3D visualization, we already had to do some small hacks to our local MNE installations for proper visualization in plot_alignment().

in examples/datasets/plot_opm_data.py, lines 96-98:
"# Due to the small number of sensors, there will be some leakage of activity
# to areas with low/no sensitivity occurs. Constraining the source space to
# areas we are sensitive to might be a good idea."
Clumsy sentence (by me), remove 'occur'.

in datasets_index.rst, lines 231-232:
"Thus the MEG<->Head and Head<->MRI coordinate transforms are both taken to
be identity matrices (i.e., everything is in MRI coordinates)..."
In the FIF files I sent, the MEG<->Head transform is not an identity matrix! However, it is included in the raw data FIF file as raw.info['dev_head_t']. Did you apply the transform directly to the channel locations?

With these small tweaks, everything should be good for review/merge. I'll leave it up to you if you want to include SQUID data as a comparison or not.

**Comment by larsoner** [comment]
@rzetter to make life easier in the future if you click over to the "Files changed" tab, you can leave comments directly in line with the code.

I'll try to address comments in the next couple of days

File: mne/viz/_3d.py [hunk 1]
```diff
@@ -1184,11 +1184,15 @@ def _make_tris_fan(n_vert):
 
 def _sensor_shape(coil):
     """Get the sensor shape vertices."""
-    rrs = np.empty([0, 2])
-    tris = np.empty([0, 3], int)
+    from scipy.spatial import ConvexHull
```
Comment (massich) [inline mne/viz/_3d.py]: what is the convention in placing imports inside the functions?

File: mne/viz/tests/test_3d.py [hunk 2]
```diff
@@ -141,7 +156,7 @@ def test_plot_evoked_field():
 @testing.requires_testing_data
 @requires_mayavi
 @traits_test
-def test_plot_alignment():
+def test_plot_alignment(tmpdir):
     """Test plotting of -trans.fif files and MEG sensor layouts."""
     # generate fiducials file for testing
     tempdir = _TempDir()
```
Comment (massich) [inline mne/viz/tests/test_3d.py]: do not overwrite tempdir

File: mne/viz/tests/test_3d.py [hunk 3]
```diff
@@ -238,6 +257,17 @@ def test_plot_alignment():
                    coord_frame='mri', subjects_dir=subjects_dir,
                    surfaces=['brain'], bem=sphere, show_axes=True)
 
+    # 3D coil with no defined draw (ConvexHull)
+    info_cube = pick_info(info, [0])
+    info_cube['chs'][0]['coil_type'] = 9999
+    with pytest.raises(RuntimeError, match='coil definition not found'):
+        plot_alignment(info_cube, meg='sensors', surfaces=())
+    coil_def_fname = op.join(str(tmpdir), 'temp')
```
Comment (massich) [inline mne/viz/tests/test_3d.py]: If you are using pytest tmpdir I would do this:
`coil_def_fname = tmpdir.join('temp')`

File: mne/viz/_3d.py (hunk 1 above)
Comment (larsoner) [reply mne/viz/_3d.py]: Some scipy imports are nested to improve `import mne` speed. There is a test for violations in `mne/tests/test_import_nesting.py::test_module_nesting` and the rules are listed there

File: mne/viz/tests/test_3d.py (hunk 2 above)
Comment (larsoner) [reply mne/viz/tests/test_3d.py]: I changed it to be `tempdir = str(tmpdir)`. This works around a `pytest` bug where `tmpdir` on older `pytest` is not seen properly as a string by some functions, so casting it explicitly to a string makes some things work better

**Comment by larsoner** [comment]
Updated example:

https://9433-1301584-gh.circle-artifacts.com/0/html/auto_examples/datasets/plot_opm_data.html

File: examples/datasets/plot_opm_data.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                    subject + '-5120-5120-5120-bem-sol.fif')
+fwd_fname = op.join(data_path, 'MEG', 'OPM', 'OPM_sample-fwd.fif')
+coil_def_fname = op.join(data_path, 'MEG', 'OPM', 'coil_def.dat')
+
+###############################################################################
+# Prepare data for localization
+# -----------------------------
+# First we filter and epoch the data:
+
+raw = mne.io.Raw(raw_fname, preload=True)
```
Comment (agramfort) [inline examples/datasets/plot_opm_data.py]: please use read_raw_fif

**Comment by agramfort** [comment]
beautiful ! thanks @larsoner @LauriParkkonen and @rzetter for making this happen.

**Pull request merged**